light-client/src/builder/light_client.rs

```diff
@@ -0,0 +1,213 @@
+//! DSL for building a light client [`Instance`]
+
+use tendermint::{block::Height, Hash};
+
+use crate::bail;
+use crate::builder::error::{self, Error};
+use crate::components::clock::Clock;
+use crate::components::io::{AtHeight, Io};
+use crate::components::scheduler::Scheduler;
+use crate::components::verifier::Verifier;
+use crate::light_client::{LightClient, Options};
+use crate::operations::Hasher;
+use crate::predicates::VerificationPredicates;
+use crate::state::{State, VerificationTrace};
+use crate::store::LightStore;
+use crate::supervisor::Instance;
+use crate::types::{LightBlock, PeerId, Status};
+
+#[cfg(feature = "rpc-client")]
+use {
+    crate::components::clock::SystemClock, crate::components::io::ProdIo,
+    crate::components::scheduler, crate::components::verifier::ProdVerifier,
+    crate::operations::ProdHasher, crate::predicates::ProdPredicates, std::time::Duration,
+    tendermint_rpc as rpc,
+};
+
+/// No trusted state has been set yet
+pub struct NoTrustedState;
+
+/// A trusted state has been set and validated
+pub struct HasTrustedState;
+
+/// Builder for a light client [`Instance`]
+#[must_use]
+pub struct LightClientBuilder<State> {
+    peer_id: PeerId,
+    options: Options,
+    io: Box<dyn Io>,
+    clock: Box<dyn Clock>,
+    hasher: Box<dyn Hasher>,
+    verifier: Box<dyn Verifier>,
+    scheduler: Box<dyn Scheduler>,
+    predicates: Box<dyn VerificationPredicates>,
+    light_store: Box<dyn LightStore>,
+
+    #[allow(dead_code)]
+    state: State,
+}
+
+impl<Current> LightClientBuilder<Current> {
```

Comment:
This pattern can definitely [be expressed both ways](http://cliffle.com/blog/rust-typestate/). I think it's just a matter of preference. I prefer the type param approach as it to me it expresses more clearly that the purpose of the struct stays the same, it's just in a different state. Perhaps it's just because I am more used to this pattern from Haskell or Scala, but in the end I don't feel that strongly either way. So if you feel like the wrapper structs would be more readable, I am happy to make the change.